src/Microsoft.IdentityModel.Tokens/Cache/MaximumSizeCache.cs

```diff
@@ -0,0 +1,253 @@
+        /// </summary>
+        /// <param name="key">The key of the item to be added to the cache.</param>
+        /// <param name="value">The value of the item to be added to the cache.</param>
+        /// <returns>true if valid, false otherwise.</returns>
+        protected static bool ValidateValues(TKey key, TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (value == null)
+                throw LogHelper.LogArgumentNullException(nameof(value));
+
+            return true;
+        }
+
+        /// <summary>
+        /// When the cache is at _maxCapacityPercentage, it needs to be compacted by _compactionPercentage.
+        /// This method calculates the new size of the cache after being compacted.
+        /// </summary>
+        /// <returns>The new target cache size after compaction.</returns>
+        protected int CalculateNewCacheSize()
+        {
+            // use the smaller of _map.Count and _capacity
+            int currentCount = Math.Min(_map.Count, _capacity);
+            return currentCount - (int)(currentCount * _compactionPercentage);
+        }
+
+        /// <summary>
+        /// Determine if the cache needs compaction.
+        /// </summary>
+        protected bool NeedsCompaction => (double)_map.Count / _capacity >= _maxCapacityPercentage;
+
+        #endregion
+
+        /// <summary>
+        /// Remove items from the map by the desired compaction percentage.
+        /// Since this is a simple hash-based cache that does not track the last used time of items, simply remove the desired number of items.
+        /// This should be a private method.
+        /// </summary>
+        private void CompactCache(object state)
+        {
+            // use the _capacity for the newCacheSize calculation in the case where the cache is experiencing overflow
+            var newCacheSize = CalculateNewCacheSize();
+            while (_map.Count > newCacheSize)
+            {
+                var item = _map.FirstOrDefault();
+                if (_map.TryRemove(item.Key, out var cacheItem))
+                    OnItemRemoved?.Invoke(cacheItem.Value);
+            }
+
+            // reset the _compactionState to CompactionStateNotRunning
+            _compactionState = CompactionStateNotRunning;
+        }
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        internal int Count => _map.Count;
+
+        #region FOR TESTING (INTERNAL ONLY)
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        internal virtual long EventQueueCount => 0;
+
+        /// <summary>
+        /// FOR TESTING PURPOSES ONLY.
+        /// </summary>
+        internal virtual int TaskCount => 0;
+
+        internal virtual void WaitForProcessing()
+        {
+            while (true)
+            {
+                if (_compactionState == CompactionStateNotRunning)
```

Comment:
`if (Volatile.Read(ref _compactionState) == ...)` ?

Reply:
@arkrasik, since this is the only operation performed in the while loop, there should not be any problem related to instructions being reordered by the compiler or processor so I don't think using Volatile.Read() is necessary based on my understanding.